feat(RAC): expose `--page-width` and `--visual-viewport-width` from Modal

lixiaoyan opened the pull request:
This allows us to properly place a dialog in a horizontally scrolled page (`document.scrollingElement.scrollLeft !== 0`).

```tsx
<ModalOverlay className="absolute top-0 left-0 h-(--page-height) w-(--page-width)">
  <Modal className="grid place-items-center sticky top-0 left-0 h-(--visual-viewport-height) w-(--visual-viewport-width)">
    <Dialog />
  </Modal>
</ModalOverlay>
```

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

## 📝 Test Instructions:

<!--- Include instructions to test this pull request -->

## 🧢 Your Project:

<!--- Company/project for pull request -->

lixiaoyan committed:
feat(RAC): expose `--page-width` and `--visual-viewport-width` from Modal

snowystinger reviewed (dismissed):
Seems reasonable

snowystinger commented:
I brought it up, the team judged it was just too close to release to take this in.

In addition, we were curious if you had tried using `100vw` and if so, what part didn't work for you? An example would be useful to show the use case for this.

rothsandro commented:
Using `100vw` (and likely also `--page-width`) for the modal overlay width results in a horizontal scrollbar during the animation, if the page has a vertical scrollbar (that is always visible and not just on scroll). This can be reproduced in the RAC docs:

<img width="1317" height="667" alt="modal" src="https://github.com/user-attachments/assets/ad059ae1-79c1-4d7a-9b84-689db56ef748" />

snowystinger commented:
> Using 100vw (and likely also --page-width) for the modal overlay width results in a horizontal scrollbar during the animation, if the page has a vertical scrollbar (that is always visible and not just on scroll). This can be reproduced in the RAC docs:

I only see the scrollbars the first time I close a Modal, after that it's fine. Given that it has 100vw both times, it doesn't seem like that is the culprit?

rothsandro commented:
> I only see the scrollbars the first time I close a Modal, after that it's fin

Hm, for me it happens always. The new docs page doesn't have a vertical scrollbar but after closing the dialog a vertical scrollbar appears and stays there, and a horizontal scrollbar appears during the animation. The old docs page already had a vertical scrollbar, the issue there was just a temporary horizontal scrollbar during the animation.

Adobe Spectrum does have this problem (and doesn't use 100vw).

https://github.com/user-attachments/assets/7dea0220-fffe-47ec-bf73-a62be258c8a9

lixiaoyan commented:
> In addition, we were curious if you had tried using 100vw and if so, what part didn't work for you? An example would be useful to show the use case for this.

In our project, we use this to implement the drawer component:

```tsx
<ModalOverlay className="absolute top-0 left-0 h-(--page-height) w-(--page-width)">
  <Modal className="grid justify-items-end sticky top-0 left-0 h-(--visual-viewport-height) w-(--visual-viewport-width)">
    <Dialog />
  </Modal>
</ModalOverlay>
```

If we use `100vw` instead, the drawer will exceed the viewport and have its right side clipped.

<img width="1528" height="815" alt="image" src="https://github.com/user-attachments/assets/a203ed17-ff16-47bc-9b5f-e020d1ecfcc4" />

snowystinger commented:
I'm really sorry, but I'm not understanding those pictures. Would you mind taking our Tailwind Starter https://react-aria.adobe.com/getting-started#storybook-starter-kits and modifying that to show the problem then pushing it to Github and sharing it here? That way we can run it and it'll be as close to your setup as possible.

Otherwise, I need some really thorough instructions on how to get to those images.

Some things to keep in mind. `100vw` doesn't take into account scrollbars, but `100%` will, so as long as you have `width: 100%` on all the parents leading up to the html document, it shouldn't be affected in that way.

It's worth noting that the reason for page-height and viewport-height variables was specifically due to iOS nav bars, which only affect height.

lixiaoyan commented:
After some testing, I discovered that, regardless of iOS, even when using `position: fixed` to place the `ModalOverlay`, the `Math.min(visualViewport.width, document.documentElement.clientWidth)` trick is still necessary. Prior to Chrome 136, an element with `position: fixed; inset: 0;` would extend to the scrollbar gutter area, resulting in anything on the right side being clipped.
https://bugs.chromium.org/p/chromium/issues/detail?id=1251856

<img width="208" height="194" alt="image" src="https://github.com/user-attachments/assets/75c197ee-a74c-4d67-899f-03b6bf03ae84" />

^ Screenshot of https://jsbin.com/vefumujapa/1/edit?html,css,js,output with an older version Chrome

lixiaoyan commented on packages/@react-aria/utils/src/useViewportSize.ts:
```diff
@@ -88,10 +88,19 @@ export function useViewportSize(): ViewportSize {
+      // The visual viewport width may include the scrollbar gutter width.
+      // See: https://github.com/w3c/csswg-drafts/issues/8099
+      ? Math.min(visualViewport.width * visualViewport.scale, document.documentElement.clientWidth)
+      : document.documentElement.clientWidth,
```
The visual viewport size does not include the scrollbar. Changed it from `window.innerWidth` to `document.documentElement.clientWidth` to get the same behavior.

See: https://drafts.csswg.org/cssom-view/#visualViewport

> Otherwise, return the width of the [visual viewport](https://drafts.csswg.org/cssom-view/#visual-viewport) excluding the width of any rendered vertical [classic scrollbar](https://drafts.csswg.org/css-overflow-3/#classic-scrollbars) that is fixed to the visual viewport.

lixiaoyan replied:
Split into https://github.com/adobe/react-spectrum/pull/9384

devongovett committed:
Merge branch 'main' into patch-6

devongovett reviewed (approved):
I think we're ok with this. Just one question above on the change to useViewportSize.

snowystinger reviewed (approved):
comment can be answered and we can follow it up

nwidynski commented:
@lixiaoyan A bit hacky, but how about calculating the visual viewport ourselves like this:

```ts
let style = getComputedStyle(document.documentElement);
let scrollBarStyle = style.scrollbarWidth;
let scrollWidth = document.documentElement.scrollWidth;
document.documentElement.style.scrollbarWidth = 'none';
let scrollBarWidth = document.documentElement.scrollWidth - scrollWidth;
document.documentElement.style.scrollbarWidth = scrollBarStyle;
let visualViewportWidth = window.innerWidth - scrollBarWidth;
```

Seems to work in all browsers pretty consistently and avoids this large mess of a polyfill. To make this even more reliable, while also avoiding potential flicker, we could also copy style from the documentElement onto a hidden element, then measure the scrollbar there - similar to our `getOffsetType` utils.

lixiaoyan commented:
https://issues.chromium.org/issues/503187943

nwidynski commented:
@lixiaoyan Yes, I understand what the issue is about. Thanks for filing it btw.

> Without this behavior change, there is no reliable way to get the viewport width excluding the scrollbar gutter in Chrome.

What I don't understand is why we can't use `window.innerWidth - scrollbarGutterWidth` as a polyfill for `document.documentElement.clientWidth` in the meantime.

lixiaoyan commented:
@nwidynski We can, but we still have no reliable way to determine the scrollbar gutter width either. From the library's perspective, the user may override the scrollbar style on the `<html>` element. We've already included something similar in our app code, but that approach isn't suitable for React Aria.
